source/compiler/qsc_qasm/src/compiler.rs

```diff
@@ -916,9 +1004,58 @@ impl QasmCompiler {
         None
     }
 
-    fn compile_pragma_stmt(&mut self, stmt: &semast::Pragma) -> Option<qsast::Stmt> {
-        self.push_unimplemented_error_message("pragma statements", stmt.span);
-        None
+    fn compile_pragma_stmt(&mut self, stmt: &semast::Pragma) {
+        fn is_parameterless_and_returns_void(args: &Arc<[Type]>, return_ty: &Arc<Type>) -> bool {
+            args.is_empty() && matches!(&**return_ty, crate::semantic::types::Type::Void)
+        }
+
+        let name_str = stmt.identifier.as_string();
+
+        // Check if the pragma is supported by the compiler.
+        // If not, we push an error message and return.
+        if !self.pragma_config.is_supported(&name_str) {
+            self.push_unsupported_error_message(format!("pragma statement: {name_str}"), stmt.span);
+            return;
+        }
+
+        // The pragma is supported, so we get the pragma kind.
+        let pragma = PragmaKind::from_str(&name_str).expect("valid pragma");
+
+        match (pragma, stmt.value.as_ref()) {
+            (PragmaKind::QdkBoxOpen, Some(value)) => {
+                if let Some(symbol) = self.symbols.get_symbol_by_name(value) {
+                    if let crate::semantic::types::Type::Function(args, return_ty) = &symbol.1.ty {
+                        if is_parameterless_and_returns_void(args, return_ty) {
+                            self.pragma_config
+                                .insert(PragmaKind::QdkBoxOpen, value.clone());
+                            return;
+                        }
+                    }
+                }
+                self.push_compiler_error(CompilerErrorKind::InvalidBoxPragmaTarget(
+                    value.to_string(),
+                    stmt.value_span.unwrap_or(stmt.span),
+                ));
+            }
+            (PragmaKind::QdkBoxOpen | PragmaKind::QdkBoxClose, None) => {
+                self.push_compiler_error(CompilerErrorKind::MissingBoxPragmaTarget(stmt.span));
+            }
```

Comment:
nitpick: could you move the error case to the top, that way the two "good" cases are together? I find that makes it easier to read this sort of pattern.